extensions/iceberg/src/main/java/io/deephaven/iceberg/layout/IcebergKeyValuePartitionedLayout.java

```diff
@@ -53,32 +53,23 @@ public IcebergKeyValuePartitionedLayout(
 
         // We can assume due to upstream validation that there are no duplicate names (after renaming) that are included
         // in the output definition, so we can ignore duplicates.
-        final MutableInt icebergIndex = new MutableInt(0);
-        final Map<String, Integer> availablePartitioningColumns = partitionSpec.fields().stream()
-                .peek(partitionField -> {
-                    // TODO (deephaven-core#6438): Add support to handle non-identity transforms
-                    if (!partitionField.transform().isIdentity()) {
-                        throw new TableDataException("Partition field " + partitionField.name() + " has a " +
-                                "non-identity transform: " + partitionField.transform() + ", which is not supported");
-                    }
-                })
+        final AtomicInteger icebergIndex = new AtomicInteger(0);
+        // TODO (deephaven-core#6438): Add better support for handling non-identity transforms
+        identityPartitioningColumns = partitionSpec.fields().stream()
+                .filter(partitionField -> partitionField.transform().isIdentity())
                 .map(PartitionField::name)
-                .map(name -> instructions.columnRenames().getOrDefault(name, name))
-                .collect(Collectors.toMap(
-                        name -> name,
-                        name -> icebergIndex.getAndIncrement(),
-                        (v1, v2) -> v1,
-                        LinkedHashMap::new));
-
-        outputPartitioningColumns = tableDef.getColumnStream()
-                .map((final ColumnDefinition<?> columnDef) -> {
-                    final Integer index = availablePartitioningColumns.get(columnDef.getName());
-                    if (index == null) {
-                        return null;
+                .map(icebergColName -> {
+                    final String dhColName = instructions.columnRenames().getOrDefault(icebergColName, icebergColName);
+                    final ColumnDefinition<?> columnDef = tableDef.getColumn(dhColName);
+                    if (tableDef.getColumn(dhColName) == null) {
+                        throw new TableDataException("Partitioning column " + dhColName + " not found in table " +
+                                "definition but corresponding identity partitioning column " + icebergColName + " is " +
+                                "present in the partition spec, table definition: " + tableDef + ", partition spec: " +
+                                partitionSpec);
                     }
-                    return new ColumnData(columnDef.getName(), TypeUtils.getBoxedType(columnDef.getDataType()), index);
+                    return new IdentityPartitioningColData(dhColName, TypeUtils.getBoxedType(columnDef.getDataType()),
+                            icebergIndex.getAndIncrement());
```

Comment:
This increment is not correct anymore because we now have a filter on the stream.